crypto-sodium/lib/Crypto/Sodium/Salt.hs

```diff
@@ -0,0 +1,73 @@
+-- SPDX-FileCopyrightText: 2021 Serokell
+--
+-- SPDX-License-Identifier: MPL-2.0
+
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | This module gives different ways of obtaining salts.
+--
+-- A “salt” is additional input provided to certain functions.
+-- Unlike a nonce, salt is used for “namespacing”, so there is not
+-- requirement that it is used only once, instead you use different
+-- salts to deisgnate different applications of an algorithm.
+--
+-- This is a subtle and pure semantical difference, so, for convenience,
+-- we also re-export some functions from "Crypto.Sodium.Nonce".
+module Crypto.Sodium.Salt
+  (
+  -- * Literals
+    utf8
+
+  -- * Random salt generation
+  , Crypto.Sodium.Nonce.generate
+  ) where
+
+import Data.ByteArray.Sized (SizedByteArray, unsafeSizedByteArray)
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Internal as BS
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import Language.Haskell.TH (Exp, Q)
+import qualified Language.Haskell.TH.Lib as TH
+import Language.Haskell.TH.Quote (QuasiQuoter (..))
+import qualified Language.Haskell.TH.Syntax as TH
+
+import qualified Crypto.Sodium.Nonce
+
+
+-- | Make a /sized/ 'ByteString' from a type-level string literal.
+--
+-- @
+-- {-# LANGUAGE QuasiQuotes #-}
+--
+-- {- ... -}
+--
+-- let salt = [utf8|hello world|] :: 'SizedByteArray' 11 ByteString
+-- @
+--
+-- This uses Template Haskell to compute the length of the UTF-8 encoding
+-- of the string you provide. Note that the string will be taken as is,
+-- with all whitespace preserved, for example @[utf8| x |]@ will have length 3.
+utf8 :: QuasiQuoter
+utf8 = QuasiQuoter { quoteExp, quotePat, quoteType, quoteDec }
+  where
+    quoteExp :: String -> Q Exp
+    quoteExp str =
+        let
+          valExpr = [e|BS.packBytes $(TH.lift $ BS.unpackBytes bs)|]
+
+          len = BS.length bs
+          t = [t|SizedByteArray $(TH.litT . TH.numTyLit . fromIntegral $ len) ByteString|]
+        in
+          -- This is really safe because, well, we compute the length.
+          [e|unsafeSizedByteArray $(valExpr) :: $(t)|]
```

Comment:
I believe the consensus is to use `[|...|]` expression quotes, I don't remember the last time I've seen `[e|...|]` in the wild.

Also, a nitpick, but you don't need `()` in single-identifier antiquotes.

```suggestion
          [|unsafeSizedByteArray $valExpr :: $t|]
```

Reply:
I guess you are right, I’ll remove the parens.

However, as for `[e|...]`, I, um... don’t think removing it is a good idea and helps readability. If there is really such consensus, then I think it is wrong :). Also, thanks, Vlad: https://hackage-search.serokell.io/?q=%5C%5Be%5C%7C

Reply:
I disagree with `e` helping readability. I won't press the issue, since you're the maintainer, and are free to choose the code style, but I will explain my rationale. Also of note that syntax highlighting on GitHub does handle `[|...|]` quotes, but not `[e|...|]` quotes (full disclosure: I'm the maintainer of the Haskell highlighter GitHub is using).

So, the quote without `e` is always the TH expression quote. It is all that it can be -- this syntax is unambiguous. However, `[e|...|]` is ambiguous: without `TemplateHaskell`/`TemplateHaskellQuotes`, but with `QuasiQuotes`, this may be an arbitrary user-defined quote. Note that this applies to typed quotes also, since `[e||...||]` is still syntactically a valid user-defined quasiquote. Hence, my point is `[|...|]` and `[||...||]` are at the end of the day more readable, because they are unambiguous.